### PR TITLE
doc/syntax: improve coding style

### PR DESCRIPTION
**Pull request by ret2libc**
I think we should start to accept patches after the code style has been followed. Otherwise we will have (like we already have actually) hard to read code, without any consistency.

### CODE REVIEW

**Comment by radare** [comment]
A tab is just 1 char, not 8. User can configure this in his editor for his own enjoyment. But the file must contain just \t

Clang format doesnt works well in all cases, also is important to note that some clang format options used by the config file i wrote depend on the very latest llvm release

**Comment by radare** [comment]
If 0 to comment code is important rule, why remove it? Its better not to conment code, but sometimes deleting it is not an option

**Comment by radare** [comment]
Dont use do statments without brackets. Ur change looks bad to me

**Comment by radare** [comment]
Use 78 columns, not 80. (If possible) some terminals fuckup the wrapping if using 80 or 79.

The only changethat makes sense to me in this patch is just the switch one

**Comment by radare** [comment]
And maybe the do without brackets can be discussed. But we should use a syntax that clangformat can support

**Comment by sushant94** [comment]
Just my comment on this (maybe not everyone likes this).
* Can we make braces compulsory for even 1 line if's etc? imo, it makes the code clearer and much more readable (and lesser room to make mistakes during refactors).

* Second point, can we make this a markdown file and put the examples inside a code block, such as:
  ```C
  if (a == b) {
      ...
  } 
```

* For comment blocks, I feel the following formatting is more readable.
    ```C
    /* 
     * This is the first line of comment that describes the comment in short.
     * Description/Body of the comment.
     * ..... (more comments)
     * Last line of comment.
     */
   ```

* Also, a mention about using early return instead of having if-else would be a good addition for people contributing for the first time. Example,
```C
// Good.
int check_something (int n) {
	if (n == 0) {
		return -1;
	}
	// Do something else
}
// Bad
int check_something_else (int n) {
	if (n != 0) {
		// Do something.
	} else {
		return -1;
	}
}
```

**Comment by radare** [comment]
> On 05 Aug 2015, at 07:32, Sushant Dinesh <notifications@github.com> wrote:
> 
> Just my comment on this (maybe not everyone likes this).
> 
> Can we make braces compulsory for even 1 line if's etc? imo, it makes the code clearer and much more readable (and lesser room to make mistakes during refactors).
> 
This rule comes from perl and the reason is readability. The older i become the more i like this rule. But sometimes i feel young and brave and i think that onelineres ifs without brackets are fine. Mainly when grouped in blocks because indentation by columns makes it more readable than multiple line with brackets (horizontal reading).

> Second point, can we make this a markdown file and put the examples inside a code block, such as:
> 
> if (a == b) {
>     ...
> } 

Agree. Rename to syntax.md
> 
> For comment blocks, I feel the following formatting is more readable.
> 
> /* 
>  * This is the first line of comment that describes the comment in short.
>  * Description/Body of the comment.
>  * ..... (more comments)
>  * Last line of comment.
>  */

Looks like Java to me. Never felt the need to add short and then long comment in the same comment. Comments are redundant information because code is what really describes what programs do. Adding a comment is dangerous because it can get unsync if code changes. And even more if you add this kind of comment. I disagree
> 
> Also, a mention about using early return instead of having if-else would be a good addition for people contributing for the first time. Example,
> 

Yep. I have also read recently a doc about swift explaining qhen to use int or unsigned int32 and such. I think this is good stuff to discuss here too.

Avoiding nested or big conditional blocks makes the code essier to read and less tree-like. But sometimes it makes the function size longer, or having multiple return points which is also bad. Because it can lead to memleaks or other issues. Need to discuss more on this

**Comment by ret2libc** [comment]
@radare tabs are 8 chars, it's important to said that, like the Linux Kernel Coding Style says (I took that from there). The reason is that if you can set tab to whatever you want, like 1 char, then you can easily put 10 level of indentation in 80 chars.

Probably I wasn't clear in the text, but do-while statements always need braces.

I've just taken a subset of the Linux Kernel Coding Style, I think it's quite clear and good. They also use 80 chars, why should we use 78 chars? Who has a screen with max 80 columns? I don't know of any after-80s terminal that doesn't work when wrapping with 80 columns.

**Comment by sushant94** [comment]
@radare comments are mainly to address the __why__ part of the code. Most of the times though __how__ is pretty clear from the code, it may not be apparent immediately as to why something is being done the way it's being done. Comments help a ton in these cases as it may explain a part of the design etc. and can quickly get a new contributor up to speed (and maintainability). It could also help people not experienced in the domain get up and running quickly. So I don't feel they are redundant.

If we have people reviewing changes (we should imo), it should not be too hard to catch the comments and code out-of-sync case.

**Comment by ret2libc** [comment]
I agree with sushant94, comments should explain why or the high level part of a complex algorithm. So they are useful sometimes (not always)

**Comment by ret2libc** [comment]
Also, clang-format is nice to use and we should suggest to use it. If you really can't because it doesn't work for you, you can always follow this coding style and common sense and someone will review the PR and tell you about the changes you have to make if something is wrong.

**Comment by ret2libc** [comment]
About multiple return points and early return, the Linux kernel coding style has a good paragraph explaining how to do that with goto. It would be probably nice to include that in our style.

**Comment by radare** [comment]
Yep. I tried 3 different formatters and clang is the ine that makes more sense. They are adding more options on every release. But iirc the last reason not to ise it is because of the ugly wrapping of long lines. Because its indenting one word per line at column 70 and such. Instead of using common sense.

I would suggest adding a target in rules.mk and slowly run 'make indent' on every module (libr/*) reviewing the changes to confirm clang format does the right thing.

Another reason not to use clang format evrywhere is because updaing codebases from external sources. Like gnu disassemblers, grub, tcc, etc will make it impossible to detect the correct patches to apply. So this should be only used in r2-only code

**Comment by radare** [comment]
That depends a bit on the stability of the code and the man/loc ratio. I prefer self explanatory code. Using meaningful macros, enums, function names, etc

**Comment by radare** [comment]
All xterms open by default at 80x24 or 80x25. The reson for 78 columns is because its 80-2, that 2 are the chars needed to draw a box around the window, usually, line number, scrollbar or stiff like that. This is inherit from mados's edit.com. But imho takes sense in modern systems when you split the terminal in multiple columns. Using wider configurations can be problematic if trying to visualize more than one file in a single screen. Also, i use to read code in my phone, which is usually less than 80 columns. But setting up to 78 fits with a pretty reasonable font size.

About tabs, this is right. I use 8 visual chars for a tab. Imho 8 is ok for C, for for pther langs like js, 2 or 4 is better.

But tab is 1 physical char. The editor can be configured to visualize those tabs as N chars up to the user preferences. But i agree that asuming 8 is good to avoid too much nesting (or quickly detect it with the eye)

**Comment by ret2libc** [comment]
I think it's not necessary to add any rule, because it's a task that we just need to do one time and also it's not an automatic thing, so we can just run clang-format on small parts of the project and integrate it slowly.

This codestyle is only for C and r2-only code. That is a big part of the project.
I agree that sometimes the wrapping of long lines is not the best, so maybe we can say that contributors should run clang-format(on r2/C code) and we can remove the column limit from the clang-format file, but just review it manually (and of course we are already telling that max columns should be 78/80).

**Comment by ret2libc** [comment]
If you read code from phone, that is usually less than 80 cols, it's also usually less than 78 chars imo. I also usually have two files on the same screen and having 78 cols or 80 cols doesn't change anything.

If you want to have line numbers and other things in the 80 cols of xterm that's another problem. Because for example, line numbers can also take 4 chars if you have 1000 lines of code. Should we use 74 chars for that reason? I don't think so, imho. 80 chars is already a small number, let's try to not reduce it too much because we want to be compatible with very old style things. I hope nobody really write code in a 80x25 box, because you can't even see a small function in that space (if you have powerline, numbers, cursor on the left, etc etc that space is also much smaller)

**Comment by crowell** [comment]
i really dislike the "no braces on if/else if it is only one line" thing.

This isn't really great for readability or maintainability.
I'd prefer to ALWAYS have braces even on one line statements.

On Wed, Aug 5, 2015 at 10:28 AM, Riccardo Schirone <notifications@github.com
> wrote:

> If you read code from phone, that is usually less than 80 cols, it's also
> usually less than 78 chars imo. I also usually have two files on the same
> screen and having 78 cols or 80 cols doesn't change anything.
>
> If you want to have line numbers and other things in the 80 cols of xterm
> that's another problem. Because for example, line numbers can also take 4
> chars if you have 1000 lines of code. Should we use 74 chars for that
> reason? I don't think so, imho. 80 chars is already a small number, let's
> try to not reduce it too much because we want to be compatible with very
> old style things. I hope nobody really write code in a 80x25 box, because
> you can't even see a small function in that space (if you have powerline,
> numbers, cursor on the left, etc etc that space is also much smaller)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/radare/radare2/pull/3035#issuecomment-128015478>.
>



-- 
Jeff Crowell
https://github.com/crowell

**Comment by ret2libc** [comment]
@crowell what about error condition checks?

```
if (!ok)
        return R_FALSE;
```

we will end up with a lot of
```
if (!ok) {
        return R_FALSE;
}
```

**Comment by crowell** [comment]
@ret2libc in this case, `if (!ok) return R_FALSE;` seems like it is ok for
me.

I don't mind same line, but I do dislike next line without braces.

On Wed, Aug 5, 2015 at 10:41 AM, Riccardo Schirone <notifications@github.com>
wrote:
>
> @crowell what about error condition checks?
>
> if (!ok)
>         return R_FALSE;
>
> we will end up with a lot of
>
> if (!ok) {
>         return R_FALSE;
> }
>
> —
> Reply to this email directly or view it on GitHub.




--
Jeff Crowell
https://github.com/crowell

**Comment by ret2libc** [comment]
+1 @crowell 
I like this way too. It can be a good idea to have always braces except for short if-statement that can fit in a one-line and are used for error condition or stuff like that.

So, it's bad to have:
```
if (ok) x+= 3;
```
We should have
```
if (ok) {
        x += 3;
}
```

But it's ok to have:
```
if (!ok) return R_FALSE;
```
or
```
if (!ok) goto err_foo;
```
@radare what do you think about it?

**Comment by radare** [comment]
I know linus is a happy goto fan. I also use some gotos because it greatly simplifies some flows but its better not to abuse them. Imho

Do you have a link for that linux-goto doc?

**Comment by ret2libc** [comment]
Chapter 7: Centralized exiting of functions of https://www.kernel.org/doc/Documentation/CodingStyle

I don't like goto, but I really think they are great for cleaning on errors. It actually makes the code easier to maintain because you have only one place where you free things.

**Comment by condret** [comment]
pls, don't become code-style-nazis

**Comment by ret2libc** [comment]
@condret we won't, but we *really* need a coding style and one that is applied everywhere. At the moment you can find in the code things like:
```
if(a==b) { return; }
if (a ==b){
...
}
else { sada }

for(i=0;i<3; i++) {
...
}

a=b + 3;
b = 5* 7;
```
in the same file!! This is not good at all. It's far from being readable code and for that we need consistency.

**Comment by Hi-Angel** [comment]
>I don't mind same line, but I do dislike next line without braces.

Some time before I was thinking like @crowell. But once I came to Haskell *(the lang takes advantage of indentation to show a code semantic, and doesn't even use braces in arguments)*, and that completely changed my mind.

Perhaps [Haskell made me worse programmer?](http://lukeplant.me.uk/blog/posts/why-learning-haskell-python-makes-you-a-worse-programmer/)

**Comment by Hi-Angel** [comment]
Anyway, why not just take an existing standard. E.g. [the Google one](https://google-styleguide.googlecode.com/svn/trunk/cppguide.html#Conditionals) — I heard there're smart folks working ;)

TL;DR: they have all the three variants, i.e.:

	if (foo) bar();
	…
	if (foo)
		bar();
	…
	if (foo) {
		bar();
	}

The one line statement is not allowed when there is an `else` statement; and missing braces are disallowed when there's an according `else` and either of them uses braces.

**Comment by ret2libc** [comment]
@Hi-Angel there are a *lot* of stuff in that standards and a lot of C++ related stuff. Moreover, they use spaces instead of tabs. I also think we don't want our contributors to have to read all those things, just a guide that would make them clear how the code should be written.

**Comment by radare** [comment]
Did u tried the clang format script provided in r2?

**Comment by radare** [comment]
Remember the goto fail? ;D

**Comment by ret2libc** [comment]
@radare yes, in general it looks good. In the last commit I just removed the check on the width of the line. I think it's a good starting point for new contributors, then we can review PR and if there's something that clang-format is not good at, we can change it manually.

What is the goto fail?

**Comment by ret2libc** [comment]
@radare also the main difference between clang-format and our current style is multiple lines.
Current:
```
a = my_super_function(a, b, c,
                d, e, f);
```
Clang-format:
```
a = my_super_function(a, b, c,
                      d, e, f);
```
I like the second one, but both are good imo.

**Comment by Hi-Angel** [comment]
>@crowell what about error condition checks?
>
>if (!ok)
>        return R_FALSE;
>
>we will end up with a lot of
>
>if (!ok) {
>        return R_FALSE;
>}

Why not use ternary operator for that purpose? It is greatly shorten the code, e.g. 

	ErrType ret = (!ok)? R_FALSE
		: (someThing2)? R_SOMETHING2
		: (someThing3)? R_SOMETHING3
		: R_DEFAULT;
	return ret;

**Comment by jjdredd** [comment]
@Hi-Angel, I hope you're kidding.

I agree with @crowell on braces. I also agree with @condret.

Imho it's silly to argue because of two additional columns of code. Although I agree on
78/80 column rule, I think that with 8-space wide indentation 80 columns are not enough.
Code becomes unreadable after 4-th level of indentation.

I think that commenting on code must be encouraged. I'd rather read a 2-line comment than
read 50 lines of code, trying to figure out what exactly does this function and what are its arguments.
Also, not everyone who contributed/will contribute to r2 can write readable code.
Sometimes you just need a general idea/design of a subsystem, or to find some part of code.
Comments help in this case. Besides comments can also prevent contributors from
"fixing" code that works, but doesn't look so obvious.

**Comment by Hi-Angel** [comment]
>@Hi-Angel, I hope you're kidding.

@jjdredd I'm not. Why?

**Comment by radare** [comment]
http://avandeursen.com/2014/02/22/gotofail-security/

**Comment by radare** [comment]
Both are bad imho. Second line should be indented one tab. Not right after the parenthesis of the above line. It just makes it unreadable when functionnname is long . If you find how to make clang format do this it will be great.

**Comment by radare** [comment]
Ternary operator for this case is even more confusing. I would suggwst using it only for two value assignements. The rest of cases is just hacky. Maybe cool to write, but harder to read

**Comment by Hi-Angel** [comment]
>Ternary operator for this case is even more confusing. I would suggwst using it only for two value assignements. The rest of cases is just hacky. Maybe cool to write, but harder to read

I'd argue that the reading is just a matter of habit. See: it is shorten the code without losing any part of sense *(the only thing it does is removing a duplicate code that appears with "if"s)*. There's just no other reason except of a habit.

I could even show a real example to convince you; I hope you won't blame me because it is C# — that's just an example.

	string error = (packtsz == 0) ? "размер пакета"
		: (delay == 0)            ? "ожидание ответа на пакет"
		: (maxfails == 0)         ? "кол-во разрешенных неудач пересылки пакета"
		: String.Empty;
	if (error != String.Empty)
	{
		MessageBox.Show("Введите " + error + "!");
		return null;
	}

Just notice: once you looked at the assignement, it becomes obvious, that the next *if*-statement is only check for an error. Without the ternary that would look like:

	if (packtsz == 0)
	{
		MessageBox.Show("Введите размер пакета");
		return null;
	}
	if (delay == 0)
	{
		MessageBox.Show("Введите ожидание ответа на пакет");
		return null;
	}
	if (maxfails == 0)
	{
		MessageBox.Show("Введите кол-во разрешенных неудач пересылки пакета");
		return null;
	}

There one should look inside every *if*-statement to be sure that it's just an error checking. What if there's even more checks? And what if that's exactly the function that one debugging? In that case every time one want to check some idea, they should again look through the blocks to find where the error-checking-block ends, either where it starts. While with the first variant it becomes immediately obvious — the ternary "highlighted" with indentation, and after that just a single "if" goes.

**Comment by ret2libc** [comment]
@Hi-Angel you are doing it differently.
```
if (packtsz == 0)
   string = "размер пакета";
else if (delay == 0) 
   string = "ожидание ответа на пакет";
else if (maxfails == 0)
   string = "кол-во разрешенных неудач пересылки пакета";
else
   string = "";

if (string != "")
    MessageBox.show(string);
```

I agree that ternary operator is just hacky for most of the cases.

**Comment by ret2libc** [comment]
@jjdredd Btw, it's good that after 4 level of indentation the code becomes unreadable, because you shouldn't have such a nested function. You should split it.

**Comment by jjdredd** [comment]
@ret2libc sometimes it's hard to split a function without passing around 10x arguments.

**Comment by ret2libc** [comment]
@jjdredd in that case you group params in a structure

**Comment by jjdredd** [comment]
@ret2libc structure that is used only once isn't an elegant solution.

**Comment by Hi-Angel** [comment]
>@Hi-Angel you are doing it differently.

@ret2libc funny, I just out of curiosity counted — your variant with the 4 lines missing lines being added *(the return statement in "if", and the declaration)* takes exactly as much space as the one you referred *(15 lines)*. But yeah, it's a bit more obvious than mine *if-statement* wall.

**Comment by jjdredd** [comment]
@Hi-Angel, I think @ret2libc meant the semantics.

**Comment by Hi-Angel** [comment]
@jjdredd the semantic most obvious in the ternary-wall — I told arguments for it, and I didn't see even a single versus. @ret2libc 's example was to show the right way to write my second example.

**Comment by jjdredd** [comment]
@Hi-Angel it's not, just because it isn't readable. That's the only versus I need.

**Comment by Hi-Angel** [comment]
@jjdredd one'd say «unreadable» for anything new. If one don't know LISP syntax, then after they look at it, they'd say the same «Unreadable». But once they learn LISP, the thought fades away. Needs I say more to that it's just a matter of a habit?

**Comment by jjdredd** [comment]
First of all some constructs are a lot easier to read habit or not. Second, the code should be clear
for people without this rare 'habit' you're talking about. Third, I am not looking and I don't think that the core devs are looking for new habits here.

**Comment by Hi-Angel** [comment]
>First of all some constructs are a lot easier to read habit or not. 

@jjdredd yeah, I do agree, there's an unreadability that comes from a syntax being overburdened — the most appealing example is C++ templates. But in the ternary construction no such a burden *(point me, if I'm wrong)*.

>Second, the code should be clear for people without this rare 'habit' you're talking about

Without the habit to the language one is using? o.O

>Third, I am not looking and I don't think that the core devs are looking for new habits here.

Well, I'm not forcing to use the syntax — the first comment when I mentioned it was just as a solution to a problem. There're others, of course.

**Comment by jjdredd** [comment]
> Without the habit to the language one is using?

No, habit of using exotic multi ternary ofc.

**Comment by jjdredd** [comment]
Perhaps add a shell script for clang-format for convenience?

**Comment by radare** [comment]
Theres already a make target for this but clang format is retarded in some cases. Cant still be used in production software imho

**Comment by ret2libc** [comment]
Ok, I changed some other things in the syntax.md. Have a look at it and tell me what you think. It seems more or less what we already have. I've also changed the README to point to the doc/syntax.md file

**Comment by radare** [comment]
please squash the 6 commits for proper review

**Comment by ret2libc** [comment]
You can squash them when you will merge them. What else do you need here?

**Comment by radare** [comment]
im against the use of multiline comments, they cant be nested and tend to be problematic i prefer single line comments. I have removed that line and changed the `if()\n\tfoo\nelse\tbar` construction. I have added a note about global variables.

See my commit after that merge. thanks!

**Comment by ret2libc** [comment]
The explanation of if/else is not consistent anymore with what you changed. 
I don't actually like the
```
if (cond) do_something();
else do_something_else();
```

but it may be ok IF both the branches (if and else) can fit in one line, otherwise it would be better to use the complete form. If only one of the two branches fits in a one-line i think it's not a good pattern.
```
if (cond) {
         do_something();
} else {
         do_something_else();
}
```

**Comment by radare** [comment]
yes i prefer the fork with brackets. but for oneliners its fine in the way i did. also, i tend to do stuff like this:

```
if (foo) {
  ...
} else eprintf ("lalala\n");
```

this way the else doesnt uses brackets and its clear what it's for. feel free to commit the changes you see in a new pr

**Comment by sushant94** [comment]
@radare I think we should make `else` bracket mandatory in case `if` has more than one-line (hence has a bracket).
This makes it more readable, clear and less error prone while refactoring that code.

**Comment by ret2libc** [comment]
+1 for @sushant94 

I actually agree with sushant... it's really not a good pattern for the mind the way it is now.
My mind doesn't know where it should look for the else instructions... with the bracket you know exactly where you will find them in case there is an else.
